### PR TITLE
Add environment-safe deploys and deployment health checks

### LINKED ISSUE #345: Enforce Firebase App Check on callable functions
Part of #326. Split from #286 — this covers the App Check half; per-UID throttling is tracked separately in #344.

> **Scope note (17 July 2026):** This issue is intentionally deferred from #365 and remains tracked independently because completion requires Firebase/Google Cloud console configuration before callable enforcement can be enabled safely.

## Problem
No App Check attestation is enforced on any callable function. A script that extracts this app's Firebase config (publicly visible in the client bundle — that's normal/expected for Firebase) can call any `onCall` function directly (e.g. via curl/Postman with a valid ID token) without ever loading the real app, bypassing whatever UI-level protections exist.

## What App Check does and doesn't cover
App Check verifies the request came from an unmodified instance of this app (via reCAPTCHA v3/Enterprise attestation), stopping bots/scrapers/replayed-config abuse that never touches the real app. It does **not** cap what a legitimate signed-in user can do through the real app UI — that's what #344 (per-UID throttling) is for. The two are complementary.

Does not apply to (and shouldn't be enabled on) the `onRequest` webhook-style endpoints — `stripeWebhookPayments`/`stripeWebhook`, `notifyDeliveryCallback`, `unsubscribeAnnouncement` — since those are hit by Stripe/GOV Notify/email clients, never by this app's own client SDK, and `enforceAppCheck` isn't even an available option on `onRequest` functions (confirmed in the `firebase-functions` v2 type definitions — only `CallableOptions`, used by `onCall`, has it).

## Blocker: requires action outside this repo
Enabling this requires:
1. Registering a reCAPTCHA v3 (or Enterprise) site key in Google Cloud Console for this project — an external step that has to be done by someone with access to the GCP project, not something completable from a code change alone.
2. Enabling App Check in the Firebase console and registering the web app.
3. Adding the site key as a client-side config value (e.g. `VITE_RECAPTCHA_SITE_KEY`).

Once the key exists, the code side is small: `initializeAppCheck()` on the client, `enforceAppCheck: true` added to each `onCall` function's options.

## Suggested approach
1. Whoever has GCP console access registers the reCAPTCHA key and enables App Check enforcement for the web app.
2. Add client-side `initializeAppCheck` (likely in `src/config/firebase.ts`) reading the site key from a new env var.
3. Add `enforceAppCheck: true` to each `onCall` function's options across `functions/src/*.ts` (not the `onRequest` webhook endpoints listed above).
4. Roll out in "monitor" mode first if Firebase supports it, to check for false positives (e.g. App Check debug tokens needed for local dev/emulator) before hard-enforcing.

### LINKED ISSUE #437: [FEATURE] Add automated Firebase/GCP deployment configuration and health checks
## Description

Implement a read-only deployment verification script that uses the Firebase CLI, Google Cloud CLI, and authenticated Google/Firebase APIs where necessary to audit the deployed state of the Dev, Beta, and Prod environments.

The check should distinguish between:

1. **Configuration audit** — required APIs, Firebase products, deployed resources, IAM, secrets, regions, rules, rewrites, headers, and other expected settings exist and match version-controlled expectations.
2. **Runtime smoke checks** — deployed endpoints and representative non-destructive application flows actually work.

Expose the check through an npm script, for example:

```sh
npm run deployment:check -- --env beta --expected-sha "$(git rev-parse HEAD)"
```

The command should be safe to run locally or in CI and must not mutate the selected environment.

---

## Context

The deployment runbook defines a schema-first rollout across Data Connect, Storage, Functions, and Hosting, but most post-deployment checkpoints are currently manual. A successful `firebase deploy` only proves that the deployment command completed; it does not prove that all required APIs are enabled, resources use the intended configuration, IAM and secrets are correct, the live Hosting release contains the expected revision, or the deployed application works end to end.

Automating these checks will make Dev → Beta → Prod promotion safer, expose configuration drift, and provide an auditable CI result.

Relevant documentation:

- `docs/operations/environments-dev-beta-prod.md`
- `docs/operations/new-production-instance.md`
- `docs/operations/section-file-storage.md`
- `docs/operations/firebase-hosting-security-headers.md`

---

### Interface and safety

- [ ] Add a documented deployment-check command supporting `dev`, `beta`, and `prod` aliases from `.firebaserc`.
- [ ] Resolve the alias to an explicit project ID and verify Firebase CLI and gcloud operations target that same project before running checks.
- [ ] Ensure all checks are read-only and fail before inspection if the environment or project cannot be resolved unambiguously.
- [ ] Return a non-zero exit status when a required check fails.
- [ ] Provide concise human-readable output plus optional machine-readable JSON suitable for CI artifacts.
- [ ] Redact access tokens, secret values, and other sensitive data from output.

### Version-controlled expectations

- [ ] Define expected resources and configuration per environment without committing secrets.
- [ ] Include the expected region, Hosting site, Data Connect service/connector, Storage bucket, required APIs, expected Functions, service accounts, and intentionally public endpoints.
- [ ] Document how expectations are updated when infrastructure changes.

### API and Firebase product configuration

- [ ] Verify all required Google Cloud APIs are enabled using `gcloud services list --enabled` or an equivalent authenticated API.
- [ ] Verify the Firebase project and expected web app exist and belong to the resolved Google Cloud project.
- [ ] Verify Data Connect is initialized and the expected `sodc-web-service` service and `api` connector are deployed.
- [ ] Verify the expected Firebase Hosting site and live channel exist.
- [ ] Verify the expected Cloud Storage for Firebase bucket exists.
- [ ] Verify relevant Authentication configuration, required sign-in providers, and authorized domains where supported by the available APIs.
- [ ] Verify App Check/reCAPTCHA configuration where supported, and clearly report settings that cannot be audited programmatically.

### Functions and runtime infrastructure

- [ ] Use the Cloud Functions v2 API/CLI to verify every expected Function exists in `europe-west2`, is `ACTIVE`, uses the expected runtime and service account, and has completed reconciliation.
- [ ] Verify underlying Cloud Run services are ready and route traffic to a ready revision where applicable.
- [ ] Verify required secret resources exist and deployed Functions reference the expected secrets without reading or printing secret values.
- [ ] Detect unexpected deployed Functions or services and report them as configurable warnings or failures.
- [ ] Verify HTTP invocation/IAM policy matches the allowlist of intentionally public endpoints.

### Storage and security configuration

- [ ] Verify bucket IAM, lifecycle, and CORS settings match the repository documentation.
- [ ] Verify required service-account roles and impersonation bindings.
- [ ] Verify deployed Storage rules through ruleset metadata and/or a negative unauthenticated read/write probe.
- [ ] Fail if a bucket or Function is unintentionally public.

### Hosting and release identity

- [ ] Request the Firebase-provided Hosting URL and at least one SPA deep link, requiring a successful response.
- [ ] Verify deployed CSP, `X-Content-Type-Options`, `Referrer-Policy`, `Permissions-Policy`, and `X-Frame-Options` headers against `firebase.json`.
- [ ] Verify `/unsubscribe` is routed to the deployed Function rather than served as the SPA fallback, using a non-destructive request.
- [ ] Add a build-generated, non-sensitive deployment manifest containing at least the Git SHA, environment, and build timestamp.
- [ ] When `--expected-sha` is supplied, fetch the manifest from Hosting and fail if it does not match.

### Smoke checks and diagnostics

- [ ] Provide unauthenticated, non-destructive smoke checks that can run by default.
- [ ] Provide a documented mechanism for optional authenticated smoke checks using a dedicated test identity and securely supplied CI credentials/tokens.
- [ ] Authenticated checks cover at least sign-in, a harmless Data Connect read, and one representative callable Function without creating production business data.
- [ ] Query recent Cloud Logging entries for deployment/startup/runtime errors using a bounded time window and report actionable summaries.
- [ ] Clearly distinguish metadata/configuration success from functional smoke-test success in the final result.

### Testing and documentation

- [ ] Unit-test project resolution, CLI/API JSON parsing, comparison logic, redaction, warning/failure policy, and exit codes using fixture data without requiring cloud credentials.
- [ ] Document local prerequisites, required read-only IAM permissions, CI authentication, usage examples, expected output, and troubleshooting.
- [ ] Update the environment deployment runbook to place this command at the appropriate Data Connect, Storage, Functions, and Hosting promotion checkpoints.

### LINKED ISSUE #439: Add environment-safe Hosting build and deploy scripts
## Problem

Firebase Hosting deploys the existing `dist` directory and does not currently run a frontend build. Meanwhile, the plain `npm run build` command invokes Vite's default `production` mode and therefore loads `.env.production.local`.

This makes it easy to deploy a stale bundle or deploy a bundle containing production Firebase credentials to the dev Hosting target. `firebase deploy --project dev` selects the Hosting destination but does not rewrite the Firebase configuration already embedded in the frontend bundle.

This was observed in practice: a deployment to dev authenticated against the production Firebase project, causing valid `sodc-web` accounts and password resets to appear broken.

## Proposed change

Add explicit scripts that pair the Vite build mode with the matching Firebase project, for example:

```json
{
  "build:dev": "tsc -b && vite build --mode development",
  "build:prod": "tsc -b && vite build --mode production",
  "deploy:hosting:dev": "npm run build:dev && firebase deploy --only hosting --project dev",
  "deploy:hosting:prod": "npm run build:prod && firebase deploy --only hosting --project prod"
}
```

Include an equivalent beta/staging script using the repository's agreed Vite mode and `.env` file naming.

Update the deployment documentation to use these scripts rather than separate manual build/deploy commands.

## Acceptance criteria

- Dev, beta, and production each have an explicit Hosting deployment command.
- Each command builds with the correct Vite mode before deploying.
- A Hosting deploy cannot silently reuse a previously built `dist` directory through the documented workflow.
- Documentation explains that `firebase deploy --project ...` does not select or replace frontend environment variables.
- Add a lightweight automated check, if practical, that each deploy script pairs the intended build mode and Firebase alias.

### PR DESCRIPTION
**Pull request by jdipper**
## Summary

- add environment-specific build and Hosting deployment commands that always rebuild with the intended Vite mode
- emit a deployment manifest containing the target environment, Git SHA, and build timestamp
- add a read-only deployment health checker covering Firebase, Google Cloud, Data Connect, Cloud SQL, Functions, secrets, Storage, malware scanning, Hosting routes and headers
- document operator usage, required access, production App Check work, CI adoption, and checks that remain manual

## Why

The existing deployment workflow could publish stale output or output built for the wrong environment, and operators had no single repeatable way to verify the deployed stack. These changes make environment targeting explicit and provide a fail-closed post-deployment check suitable for local use and future CI automation.

## Validation

- `npm run lint`
- `npm run test:run` — 80 files, 632 tests passed
- `npm run test:coverage` — all coverage thresholds passed
- `npm run build:dev`
- `npm run build:beta`
- `npm run build:prod`
- `node --check scripts/deployment-check.mjs`
- `node --check scripts/deployment-check-lib.mjs`
- `git diff --check`

A read-only check against Dev also confirmed the Firebase target, project, web app, Data Connect service, Hosting site, security headers, and public routes. The Google Cloud checks correctly stopped at expired local `gcloud` authentication, and the deployment-manifest check correctly reported that the currently deployed Dev build predates this change.

## Operator note

This documents the remaining production setup for #345, including registering production App Check applications and domains, but does not close #345 or enable enforcement itself.

Closes #437
Closes #439

### CODE REVIEW

**Comment by jdipper** [comment]
## Review

**Scope:** 16 files, +1,365/-21, no `functions/` changes. Adds environment-pinned build/deploy npm scripts, a Vite-emitted deployment manifest, and a large read-only `deployment-check.mjs` health checker.

### Blocking: CI is currently red
`gh pr checks 455` shows **Functions full test suite** and **Functions coverage gate** failing (both jobs, same single test):

```
FAIL src/__tests__/deploymentRunbookContracts.test.ts > deployment runbook contracts > orders full-stack releases schema-first in every environment
AssertionError: expected -1 to be greater than 3368
❯ src/__tests__/deploymentRunbookContracts.test.ts:40:21
```

Root cause: `functions/src/__tests__/deploymentRunbookContracts.test.ts` (not touched by this PR) pins the literal string `firebase deploy --only hosting` inside the "Full-stack rollout sequence" section of `docs/operations/environments-dev-beta-prod.md`. This PR's doc changes intentionally replace that raw command with the new, safer `npm run deploy:hosting:dev` — exactly the fix this PR is making, so no more hand-run `firebase deploy --only hosting` that can reuse a stale `dist/`. But the substring the test searches for via `indexOf` no longer exists in that section, so `hosting` resolves to `-1` and `expect(hosting).toBeGreaterThan(functions)` fails. The test needs updating to check for `deploy:hosting:` (or similar) instead of the old raw command, or this PR will break `main`'s Functions suite on merge.

### What's done well (the rest of the diff)
- **`resolveTarget`** cross-checks `.firebaserc`'s alias against the expected `projectId` in `config/deployment-check.json`, failing closed on mismatch — directly addresses the "publish to the wrong environment" problem in the PR description.
- **`redact()`** in `deployment-check-lib.mjs` scrubs keys matching `authorization|bearer|password|secret|token|private.?key` and scrubs Bearer tokens / query-string secrets from string values, applied to every result before it's ever printed or written.
- **Storage and IAM checks** line up correctly with the signed-URL/malware-scanner design from the section-files work: uniform bucket access, public-access prevention, the 1-day upload-prefix lifecycle rule, exact-origin PUT CORS, the runtime service account's `storage.objectAdmin` and self `serviceAccountTokenCreator` bindings (needed for keyless V4 signing), and that the malware scanner is *not* publicly invokable while the Functions runtime *can* invoke it.
- **Hosting runtime checks** confirm `/unsubscribe` is actually routed to its Function rather than silently swallowed by the SPA catch-all rewrite, and validate the new deployment manifest's `environment`/`gitSha` against the target and an optional `--expected-sha`.
- **`hostingDeploymentScripts.test.ts`** pins the exact `build:*`/`deploy:hosting:*` script strings against `.firebaserc`, and separately asserts every `deploy:hosting:*` script starts with `npm run build:<alias> &&` — good regression protection against exactly the "deploy without rebuilding" mistake this PR fixes.
- **`--authenticated` smoke check** requires `SODC_DEPLOYMENT_CHECK_AUTH_TOKEN` from the environment (never a CLI arg), with `--help` text explicitly warning not to pass or store a token in the repo.
- Uses `execFile` with array args throughout (not shell strings), so there's no injection risk from `projectId`/config values reaching a shell.

### Verdict
The tooling itself is well-designed and matches its stated goals, but this PR cannot merge as-is — the pinned runbook contract test needs updating to match the new documented command. Everything else looks solid.

File: scripts/deployment-check.mjs [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+      };
+    },
+  });
+
+  results.push(
+    result(
+      "manual-controls",
+      RESULT_STATUS.WARN,
+      "Manually confirm App Check enforcement, Auth authorized domains/providers, deployed Storage rules, scanner IAM/definitions freshness, backups, alerts, and unexpected public invokers."
+    )
```
[the remark this answers is not in the document]
Comment (jdipper) [reply scripts/deployment-check.mjs:775]: Addressed in f5fbdba. The checker now discovers HTTP/callable contracts, requires them to match the reviewed `expectedPublicInvokerFunctions` allowlist, and inspects the underlying Cloud Run IAM policy for every deployed Gen2 Function. It fails on unexpected public invokers, expected endpoints that are unexpectedly private, unreadable policies, or missing audits. Scheduled/task Functions remain outside the public allowlist. Focused tests cover public-principal detection and each mismatch category, and both full coverage suites pass.

**Comment by jdipper** [comment]
Addressed both review findings in f5fbdba:

- Updated the Functions runbook contract test to assert the new environment-safe `npm run deploy:hosting:dev` command, fixing the failing full-suite and coverage jobs.
- Added a reviewed allowlist for all HTTP/callable Function transports and automated IAM inspection of every deployed Gen2 Function's underlying Cloud Run service. The deployment audit now fails for unexpected public invokers, expected endpoints that are unexpectedly private, unreadable policies, and missing audits.
- Updated the operational guidance and removed unexpected Function invokers from the manual-only checklist.

Validation completed locally:

- Root coverage: 80 files / 635 tests passed
- Functions coverage: 73 files / 730 tests passed
- Root and Functions lint passed
- Root and Functions builds passed
- Node syntax and `git diff --check` passed

**Comment by jdipper** [comment]
## Follow-up review

Reviewed `f5fbdba Address deployment safety review` against the CI failure from my comment above. All 9 checks now pass.

### The CI-breaking issue is fixed
`deploymentRunbookContracts.test.ts` now searches for `npm run deploy:hosting:dev` instead of the old `firebase deploy --only hosting` — exactly the one-line fix needed.

### The commit goes further than the fix
Beyond the test fix, this commit turns a manual checklist item ("unexpected public invokers") into an automated check:

- **New `function-public-contract` check**: `discoverFunctionContracts` now also classifies each exported function by trigger type (`onCall`/`onRequest` → expected to need a public invoker at the transport layer; `onSchedule`/`onTaskDispatched` → should never be public) and diffs that against a new checked-in `expectedPublicInvokerFunctions` allowlist in `config/deployment-check.json`, failing on drift in either direction.
- **New `function-cloud-run-iam` check**: for every deployed function, fetches its underlying Cloud Run service's IAM policy (bounded to 6 concurrent `gcloud` calls via a small worker-pool helper) and flags `unexpectedlyPublic` (a function that shouldn't be public but is — a real exposure bug, e.g. a scheduled function accidentally left invokable by `allUsers`), `unexpectedlyPrivate` (a callable that should be public but isn't — would break the Callable SDK), `missingAudits`, and `unreadable` policies separately.
- **Self-verifying test**: `scripts/deployment-check.test.mjs` now asserts `config/deployment-check.json`'s `expectedPublicInvokerFunctions` list exactly equals what's discovered by scanning `functions/src` — so the allowlist can't silently drift from the actual source, it's enforced at test time, not just trusted.
- **Docs updated consistently**: `deployment-checks.md` moved "unexpected Cloud Run invoker bindings" out of the manual-checklist section (since it's now automated) and added a note warning not to widen the allowlist just to make the check pass without confirming the endpoint's actual auth model.

`isPublicInvokerPolicy` and `assessFunctionInvokerPolicies` logic checks out, and their new unit tests cover the relevant edge cases correctly (service-account members don't count as "public", wrong role is ignored, all four mismatch categories are distinguished). This tool still isn't wired into any GitHub Actions workflow, consistent with the PR's "future CI automation" framing rather than a live gate.

### Verdict
The blocking issue is resolved, and the follow-up work is solid, well-tested, and consistent with the PR's stated goals. No remaining concerns from my side.
